sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/SQLHelper.scala

```diff
@@ -61,4 +66,21 @@ trait SQLHelper {
     path.delete()
     try f(path) finally Utils.deleteRecursively(path)
   }
+
+
+  def testSpecialDatetimeValues[T](test: ZoneId => T): Unit = {
+    DateTimeTestUtils.outstandingTimezonesIds.foreach { timeZone =>
+      withSQLConf(SQLConf.SESSION_LOCAL_TIMEZONE.key -> timeZone) {
+        val zoneId = getZoneId(timeZone)
+        // The test can fail around midnight if it gets the reference value
+        // before midnight but tested code resolves special value after midnight.
+        // Retry can guarantee that both values were taken on the same day.
+        try {
+          test(zoneId)
+        } catch {
+          case NonFatal(_) => test(zoneId)
```

Comment:
shall we set a threshold like 3? otherwise we may hit dead loop for real failures.

Reply:
Maybe I don't understand how such loop could happen but this is not recursive code. `test` can be invoke maximum 2 times.

Reply:
ah I mis-looked. It's not recursive. LGTM